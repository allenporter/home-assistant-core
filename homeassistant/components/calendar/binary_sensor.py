"""Binary sensor for calendar to match a specific Calendar Event."""

from __future__ import annotations

from collections.abc import Callable
import datetime
import logging
from typing import Any

from homeassistant.components.binary_sensor import BinarySensorEntity

# from homeassistant.components.sensor import PLATFORM_SCHEMA, SensorEntity
from homeassistant.config_entries import ConfigEntry
from homeassistant.const import (
    CONF_ENTITY_ID,
    CONF_EVENT,
    CONF_NAME,
    CONF_OFFSET,
    CONF_PLATFORM,
)
from homeassistant.core import CALLBACK_TYPE, Context, HomeAssistant, callback
from homeassistant.exceptions import HomeAssistantError
from homeassistant.helpers import entity_registry as er, trigger as trigger_helper
from homeassistant.helpers.entity_component import EntityComponent
from homeassistant.helpers.entity_platform import AddEntitiesCallback
from homeassistant.util import dt as dt_util

from . import CalendarEntity, CalendarEvent
from .const import CALENDAR_EVENT, CONF_SEARCH, DOMAIN, EVENT_END, EVENT_START

_LOGGER = logging.getLogger(__name__)

ICON = "mdi:calendar-text"
ZERO = datetime.timedelta(seconds=0)


async def async_setup_entry(
    hass: HomeAssistant,
    config_entry: ConfigEntry,
    async_add_entities: AddEntitiesCallback,
) -> None:
    """Initialize Calendar event config entry."""
    registry = er.async_get(hass)
    # Validate + resolve entity registry id to entity_id
    calendar_entity_id = er.async_validate_entity_id(
        registry, config_entry.options[CONF_ENTITY_ID]
    )

    def is_match(calendar_event: CalendarEvent) -> bool:
        search = config_entry.options[CONF_SEARCH].lower()
        _LOGGER.debug("searching summary %s", calendar_event.summary)
        if calendar_event.summary and search in calendar_event.summary.lower():
            return True
        if calendar_event.description and search in calendar_event.description.lower():
            return True
        return False

    calendar_event_sensor = CalendarEventSensor(
        name=config_entry.options[CONF_NAME],
        calendar_entity_id=calendar_entity_id,
        unique_id=config_entry.entry_id,
        event_predicate=is_match,
    )

    async_add_entities([calendar_event_sensor])


class CalendarEventSensor(BinarySensorEntity):
    """A sensor that matches a specific calednar event."""

    _attr_has_entity_name = True
    _attr_name: str
    _attr_icon = ICON
    _attr_should_poll = False

    def __init__(
        self,
        name: str,
        calendar_entity_id: str,
        unique_id: str,
        event_predicate: Callable[[CalendarEvent], bool],
    ) -> None:
        """Initialize the calendar event sensor."""
        self._attr_unique_id = unique_id
        self._attr_name = name.capitalize()
        self._calendar_entity_id = calendar_entity_id
        self._event_predicate = event_predicate
        self._unsub: CALLBACK_TYPE | None = None
        self._attr_is_on = False
        self._is_on_count = 0  # Tracks possibly overlapping events

    async def async_added_to_hass(self) -> None:
        """Determine if any events are active right now and register triggers.

        This will register any triggers to get notified about any calendar event
        start/end events. On startup, we also need to handle any currently
        active events to set the initial on/off state.
        """
        await super().async_added_to_hass()

        component: EntityComponent = self.hass.data[DOMAIN]
        if not (
            entity := component.get_entity(self._calendar_entity_id)
        ) or not isinstance(entity, CalendarEntity):
            raise HomeAssistantError(
                f"Entity does not exist {self.entity_id} or is not a calendar entity"
            )

        for active_event in await self._async_get_active_events(entity):
            self._async_consume_event(is_on=True, calendar_event=active_event)

        configs = [
            {
                CONF_PLATFORM: DOMAIN,
                CONF_ENTITY_ID: self._calendar_entity_id,
                CONF_EVENT: event_type,
                CONF_OFFSET: ZERO,
            }
            for event_type in (EVENT_START, EVENT_END)
        ]
        self._unsub = await trigger_helper.async_initialize_triggers(
            self.hass,
            configs,
            self._trigger_action,
            DOMAIN,
            self._attr_name,
            _LOGGER.log,
        )

    async def _async_get_active_events(
        self, entity: CalendarEntity
    ) -> list[CalendarEvent]:
        """Return any currently active events."""
        now = dt_util.utcnow()
        return await entity.async_get_events(
            self.hass, now - datetime.timedelta(minutes=1), now
        )

    @callback
    def _trigger_action(
        self, run_variables: dict[str, Any], context: Context | None = None
    ) -> None:
        trigger_data = run_variables["trigger"]
        is_on = trigger_data[CONF_EVENT] == EVENT_START
        self._async_consume_event(
            is_on, CalendarEvent.from_dict(trigger_data[CALENDAR_EVENT])
        )
        self.async_write_ha_state()

    def _async_consume_event(self, is_on: bool, calendar_event: CalendarEvent) -> None:
        """Determine if the event matches the filters and update appropriate attributes."""
        if not self._event_predicate(calendar_event):
            return
        if is_on:
            self._is_on_count += 1
        else:
            self._is_on_count = max(0, self._is_on_count - 1)
        self._attr_is_on = self._is_on_count > 0

    async def async_will_remove_from_hass(self) -> None:
        """Remove triggers."""
        await super().async_will_remove_from_hass()
        if self._unsub:
            self._unsub()
