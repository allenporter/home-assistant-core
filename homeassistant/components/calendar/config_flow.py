"""Config flow for Calendar integration.

XXX
"""
from __future__ import annotations

from collections.abc import Mapping
from typing import Any, cast

import voluptuous as vol

from homeassistant.const import CONF_ENTITY_ID, CONF_NAME
from homeassistant.helpers import selector
from homeassistant.helpers.schema_config_entry_flow import (
    SchemaConfigFlowHandler,
    SchemaFlowFormStep,
    SchemaFlowMenuStep,
)

from .const import CONF_SEARCH, DOMAIN

OPTIONS_SCHEMA = vol.Schema(
    {
        vol.Optional(CONF_SEARCH): selector.TextSelector(),
    }
)

CONFIG_SCHEMA = vol.Schema(
    {
        vol.Required(CONF_NAME): selector.TextSelector(),
        vol.Required(CONF_ENTITY_ID): selector.EntitySelector(
            selector.EntitySelectorConfig(domain=DOMAIN),
        ),
    }
).extend(OPTIONS_SCHEMA.schema)

CONFIG_FLOW: dict[str, SchemaFlowFormStep | SchemaFlowMenuStep] = {
    "user": SchemaFlowFormStep(CONFIG_SCHEMA)
}

OPTIONS_FLOW: dict[str, SchemaFlowFormStep | SchemaFlowMenuStep] = {
    "init": SchemaFlowFormStep(OPTIONS_SCHEMA)
}


class ConfigFlowHandler(SchemaConfigFlowHandler, domain=DOMAIN):
    """Handle a config or options flow for Threshold."""

    config_flow = CONFIG_FLOW
    options_flow = OPTIONS_FLOW

    def async_config_entry_title(self, options: Mapping[str, Any]) -> str:
        """Return config entry title."""
        return cast(str, options[CONF_NAME])
