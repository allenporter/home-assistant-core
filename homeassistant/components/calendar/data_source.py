"""Calendar data source."""

import dataclasses
from typing import Any

import voluptuous as vol

from homeassistant.const import CONF_ENTITY_ID, CONF_TYPE
from homeassistant.core import HomeAssistant
import homeassistant.helpers.config_validation as cv
from homeassistant.helpers.entity_component import EntityComponent
from homeassistant.helpers.typing import ConfigType

from . import DOMAIN, CalendarEntity, event_dict_factory

DATA_SOURCE_SCHEMA = vol.Schema(
    {
        vol.Required(CONF_TYPE): vol.In("get_events"),
        vol.Required(CONF_ENTITY_ID): cv.entity_id,
        vol.Required("start"): cv.datetime,
        vol.Required("end"): cv.datetime,
    }
)


async def async_get_data(hass: HomeAssistant, config: ConfigType) -> Any:
    """Provide the data for the specified data source configuration."""
    component: EntityComponent[CalendarEntity] = hass.data[DOMAIN]
    entity_id = config[CONF_ENTITY_ID]
    if not (entity := component.get_entity(entity_id)):
        raise vol.Invalid(f"Entity '{entity_id}' not found")
    events = await entity.async_get_events(hass, config["start"], config["end"])
    return [
        dataclasses.asdict(event, dict_factory=event_dict_factory) for event in events
    ]
