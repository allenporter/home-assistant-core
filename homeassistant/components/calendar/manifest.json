{
  "domain": "calendar",
  "integration_type": "helper",
  "name": "Calendar",
  "codeowners": [
    "@home-assistant/core"
  ],
  "dependencies": [
    "http"
  ],
  "documentation": "https://www.home-assistant.io/integrations/calendar",
  "integration_type": "helper",
  "quality_scale": "internal",
  "config_flow": true
}