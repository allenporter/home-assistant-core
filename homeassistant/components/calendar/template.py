"""Template platform."""

from __future__ import annotations

import asyncio
from datetime import datetime

from homeassistant.core import HomeAssistant

from . import DOMAIN, CalendarEntity


def example_call(hass: HomeAssistant) -> str:
    """Return events during a specified time."""
    return "3"


def get_events(
    hass: HomeAssistant, entity_id: str, start_date: datetime, end_date: datetime
) -> list | None:
    """Return events during a specified time."""
    return asyncio.run_coroutine_threadsafe(
        async_get_events(hass, entity_id, start_date, end_date), hass.loop
    ).result()


async def async_get_events(
    hass: HomeAssistant, entity_id: str, start_date: datetime, end_date: datetime
) -> list | None:
    """Return events during a specified time."""
    component = hass.data[DOMAIN]
    entity = component.get_entity(entity_id)
    if not isinstance(entity, CalendarEntity):
        return None
    event_list = await entity.async_get_events(hass, start_date, end_date)
    return event_list


TEMPLATE_FUNCTIONS = {
    "example_call",
    "get_events",
}
