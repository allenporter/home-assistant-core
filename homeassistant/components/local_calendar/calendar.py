"""Calendar platform for a Local Calendar."""

from __future__ import annotations

from datetime import datetime
import logging
from typing import Any

from ical.calendar import Calendar
from ical.calendar_stream import IcsCalendarStream
from ical.event import Event
from ical.store import EventStore
from ical.types import Range, Recur
import voluptuous as vol

from homeassistant.components.calendar import (
    ENTITY_ID_FORMAT,
    CalendarEntity,
    CalendarEvent,
)
from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant
from homeassistant.helpers import config_validation as cv, entity_platform
from homeassistant.helpers.entity import generate_entity_id
from homeassistant.helpers.entity_platform import AddEntitiesCallback
from homeassistant.util import dt as dt_util

from .const import CONF_CALENDAR_NAME, DOMAIN
from .store import LocalCalendarStore

_LOGGER = logging.getLogger(__name__)


EVENT_DESCRIPTION = "description"
EVENT_END = "dtend"
EVENT_START = "dtstart"
EVENT_SUMMARY = "summary"
EVENT_RECURRENCE_ID = "recurrence_id"
EVENT_RECURRENCE_RANGE = "recurrence_range"
EVENT_RRULE = "rrule"
EVENT_UID = "uid"

SERVICE_CREATE_EVENT = "create_event"
CREATE_EVENT_SCHEMA = vol.All(
    cv.make_entity_service_schema(
        {
            vol.Required(EVENT_SUMMARY): cv.string,
            vol.Required(EVENT_START): vol.Any(cv.date, cv.datetime),
            vol.Required(EVENT_END): vol.Any(cv.date, cv.datetime),
            vol.Optional(EVENT_DESCRIPTION, default=""): cv.string,
            vol.Optional(EVENT_RRULE, default=""): cv.string,
        }
    ),
)

SERVICE_DELETE_EVENT = "delete_event"
DELETE_EVENT_SCHEMA = vol.All(
    cv.make_entity_service_schema(
        {
            vol.Required(EVENT_UID): cv.string,
            vol.Optional(EVENT_RECURRENCE_ID): cv.string,
            vol.Optional(EVENT_RECURRENCE_RANGE): cv.string,
        }
    ),
)


async def async_setup_entry(
    hass: HomeAssistant,
    config_entry: ConfigEntry,
    async_add_entities: AddEntitiesCallback,
) -> None:
    """Set up the local calendar platform."""
    store = hass.data[DOMAIN][config_entry.entry_id]
    ics = await store.async_load()
    calendar = IcsCalendarStream.calendar_from_ics(ics)

    name = config_entry.data[CONF_CALENDAR_NAME]
    entity_id = generate_entity_id(ENTITY_ID_FORMAT, name, hass=hass)
    entity = LocalCalendarEntity(store, calendar, name, entity_id)
    async_add_entities([entity], True)

    platform = entity_platform.async_get_current_platform()
    platform.async_register_entity_service(
        SERVICE_CREATE_EVENT,
        CREATE_EVENT_SCHEMA,
        "async_create_event",
    )
    platform.async_register_entity_service(
        SERVICE_DELETE_EVENT,
        DELETE_EVENT_SCHEMA,
        "async_delete_event",
    )


class LocalCalendarEntity(CalendarEntity):
    """A calendar entity backed by a local iCalendar file."""

    _attr_has_entity_name = True

    def __init__(
        self, store: LocalCalendarStore, calendar: Calendar, name: str, entity_id: str
    ) -> None:
        """Initialize LocalCalendarEntity."""
        self._store = store
        self._calendar = calendar
        self._event: CalendarEvent | None = None
        self._attr_name = name.capitalize()
        self.entity_id = entity_id
        self._attr_unique_id = calendar.prodid

    @property
    def event(self) -> CalendarEvent | None:
        """Return the next upcoming event."""
        return self._event

    async def async_get_events(
        self, hass: HomeAssistant, start_date: datetime, end_date: datetime
    ) -> list[CalendarEvent]:
        """Get all events in a specific time frame."""
        events = self._calendar.timeline.overlapping(start_date, end_date)
        return [_get_calendar_event(event) for event in events]

    async def async_update(self) -> None:
        """Update entity state with the next upcoming event."""
        events = self._calendar.timeline.active_after(dt_util.now())
        if event := next(events, None):
            self._event = _get_calendar_event(event)
        else:
            self._event = None

    async def _async_store(self) -> None:
        """Persist the calendar to disk."""
        content = IcsCalendarStream.calendar_to_ics(self._calendar)
        await self._store.async_store(content)

    async def async_create_event(self, **kwargs: Any) -> None:
        """Add a new event to calendar."""
        event = Event.parse_obj(
            {
                EVENT_SUMMARY: kwargs[EVENT_SUMMARY],
                EVENT_START: kwargs[EVENT_START],
                EVENT_END: kwargs[EVENT_END],
                EVENT_DESCRIPTION: kwargs.get(EVENT_DESCRIPTION),
            }
        )
        if rrule := kwargs.get(EVENT_RRULE):
            event.rrule = Recur.from_rrule(rrule)

        EventStore(self._calendar).add(event)
        await self._async_store()

    async def async_delete_event(
        self,
        uid: str,
        recurrence_id: str | None = None,
        recurrence_range: str | None = None,
    ) -> None:
        """Cancel an event on the calendar."""
        range_value: Range = Range.NONE
        if recurrence_range == Range.THIS_AND_FUTURE:
            range_value = Range.THIS_AND_FUTURE
        EventStore(self._calendar).delete(
            uid,
            recurrence_id=recurrence_id,
            recurrence_range=range_value,
        )
        await self._async_store()


def _get_calendar_event(event: Event) -> CalendarEvent:
    """Return a CalendarEvent from an API event."""
    return CalendarEvent(
        summary=event.summary,
        start=event.start,
        end=event.end,
        description=event.description,
        location=event.location,
    )
