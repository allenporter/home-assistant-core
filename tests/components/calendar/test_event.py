"""Tests for calendar event entity."""

from __future__ import annotations

import datetime
import logging
from typing import Any

import pytest
from syrupy.assertion import SnapshotAssertion

from homeassistant.components.calendar import DOMAIN
from homeassistant.config_entries import ConfigEntryState
from homeassistant.core import HomeAssistant
from homeassistant.setup import async_setup_component

from .conftest import FakeSchedule

from tests.common import MockConfigEntry

_LOGGER = logging.getLogger(__name__)

CALENDAR_ENTITY_ID = "calendar.calendar_2"
CONFIG = {DOMAIN: {"platform": "demo"}}


@pytest.fixture(autouse=True)
async def setup_freezer(freezer):
    """Set fake times used during tests."""
    freezer.move_to("2022-08-01 10:31:02+00:00")


@pytest.fixture(autouse=True)
def set_time_zone(hass):
    """Set the time zone for the tests."""
    hass.config.set_time_zone("America/Regina")  # UTC-6


@pytest.fixture(autouse=True)
async def setup_calendar(hass: HomeAssistant, fake_schedule: FakeSchedule) -> None:
    """Initialize the demo calendar."""
    assert await async_setup_component(hass, DOMAIN, CONFIG)
    await hass.async_block_till_done()


@pytest.mark.parametrize(
    "options",
    [({}), ({"search": "trash"}), ({"search": "Trash"})],
    ids=("empty", "lower", "upper"),
)
async def test_all_day(
    hass: HomeAssistant,
    fake_schedule: FakeSchedule,
    options: dict[str, Any],
    snapshot: SnapshotAssertion,
):
    """Test event entity matching an all day event."""
    state = hass.states.get(CALENDAR_ENTITY_ID)
    assert state

    fake_schedule.create_event(
        summary="Trash day calendar event",
        start=datetime.date.fromisoformat("2022-08-08"),
        end=datetime.date.fromisoformat("2022-08-09"),
    )

    config_entry = MockConfigEntry(
        domain=DOMAIN,
        data={},
        options={
            "entity_id": CALENDAR_ENTITY_ID,
            "name": "Trash day",
            **options,
        },
    )
    config_entry.add_to_hass(hass)
    await hass.config_entries.async_setup(config_entry.entry_id)
    assert config_entry.state is ConfigEntryState.LOADED

    state = hass.states.get("event.trash_day")
    assert state
    assert state.name == "Trash day"
    assert state.state == "unknown"

    assert (
        await fake_schedule.fire_state_changes(
            "event.trash_day",
            [
                datetime.datetime.fromisoformat("2022-08-07 23:59:00-06:00"),
                datetime.datetime.fromisoformat("2022-08-08 00:00:01-06:00"),
                datetime.datetime.fromisoformat("2022-08-09 00:00:01-06:00"),
            ],
        )
        == snapshot
    )


async def test_multiple_all_day(
    hass: HomeAssistant, fake_schedule: FakeSchedule, snapshot: SnapshotAssertion
):
    """Test multiple events, some matching and some not, across a few days."""
    state = hass.states.get(CALENDAR_ENTITY_ID)
    assert state

    fake_schedule.create_event(
        summary="Trash day calendar event #1",
        start=datetime.date.fromisoformat("2022-08-08"),
        end=datetime.date.fromisoformat("2022-08-09"),
    )
    fake_schedule.create_event(
        summary="Vacation",
        start=datetime.date.fromisoformat("2022-08-12"),
        end=datetime.date.fromisoformat("2022-08-13"),
    )
    fake_schedule.create_event(
        summary="Trash day calendar event #2",
        start=datetime.date.fromisoformat("2022-08-15"),
        end=datetime.date.fromisoformat("2022-08-16"),
    )

    config_entry = MockConfigEntry(
        domain=DOMAIN,
        data={},
        options={
            "entity_id": CALENDAR_ENTITY_ID,
            "search": "Trash",
            "name": "Trash Day",
        },
    )
    config_entry.add_to_hass(hass)
    await hass.config_entries.async_setup(config_entry.entry_id)
    assert config_entry.state is ConfigEntryState.LOADED

    state = hass.states.get("event.trash_day")
    assert state
    assert state.name == "Trash day"
    assert state.state == "unknown"

    assert (
        await fake_schedule.fire_state_changes(
            "event.trash_day",
            [
                datetime.datetime.fromisoformat("2022-08-08 00:01:00-06:00"),
                datetime.datetime.fromisoformat("2022-08-09 00:00:01-06:00"),
                datetime.datetime.fromisoformat("2022-08-12 00:01:00-06:00"),
                datetime.datetime.fromisoformat("2022-08-15 00:01:00-06:00"),
            ],
        )
        == snapshot
    )


async def test_overlapping_events(
    hass: HomeAssistant, fake_schedule: FakeSchedule, snapshot: SnapshotAssertion
):
    """Test matching events that match and overlap."""
    state = hass.states.get(CALENDAR_ENTITY_ID)
    assert state

    fake_schedule.create_event(
        summary="Front light",
        start=datetime.datetime.fromisoformat("2022-08-08 18:30:00-06:00"),
        end=datetime.datetime.fromisoformat("2022-08-08 23:00:00-06:00"),
    )
    fake_schedule.create_event(
        summary="Back light",
        start=datetime.datetime.fromisoformat("2022-08-08 20:30:00-06:00"),
        end=datetime.datetime.fromisoformat("2022-08-08 22:00:00-06:00"),
    )

    config_entry = MockConfigEntry(
        domain=DOMAIN,
        data={},
        options={
            "entity_id": CALENDAR_ENTITY_ID,
            "search": "light",
            "name": "Exterior lights",
        },
    )
    config_entry.add_to_hass(hass)
    await hass.config_entries.async_setup(config_entry.entry_id)
    assert config_entry.state is ConfigEntryState.LOADED

    state = hass.states.get("event.exterior_lights")
    assert state
    assert state.name == "Exterior lights"
    assert state.state == "unknown"

    assert (
        await fake_schedule.fire_state_changes(
            "event.exterior_lights",
            [
                datetime.datetime.fromisoformat("2022-08-08 18:29:00-06:00"),
                datetime.datetime.fromisoformat("2022-08-08 18:31:00-06:00"),
                datetime.datetime.fromisoformat("2022-08-08 20:31:00-06:00"),
                datetime.datetime.fromisoformat("2022-08-08 22:01:00-06:00"),
                datetime.datetime.fromisoformat("2022-08-08 23:01:00-06:00"),
            ],
        )
        == snapshot
    )
