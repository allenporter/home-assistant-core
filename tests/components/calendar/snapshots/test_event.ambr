# serializer version: 1
# name: test_all_day
  list([
    tuple(
      'unknown',
      dict({
        'event_type': None,
        'event_types': list([
          'start',
          'end',
        ]),
        'friendly_name': 'Trash day',
        'icon': 'mdi:calendar-text',
      }),
    ),
    tuple(
      '2022-08-08T06:00:00.000+00:00',
      dict({
        'calendar_event': dict({
          'all_day': True,
          'end': '2022-08-09',
          'start': '2022-08-08',
          'summary': 'Trash day calendar event',
        }),
        'event_type': 'start',
        'event_types': list([
          'start',
          'end',
        ]),
        'friendly_name': 'Trash day',
        'icon': 'mdi:calendar-text',
      }),
    ),
    tuple(
      '2022-08-09T06:00:01.000+00:00',
      dict({
        'calendar_event': dict({
          'all_day': True,
          'end': '2022-08-09',
          'start': '2022-08-08',
          'summary': 'Trash day calendar event',
        }),
        'event_type': 'end',
        'event_types': list([
          'start',
          'end',
        ]),
        'friendly_name': 'Trash day',
        'icon': 'mdi:calendar-text',
      }),
    ),
  ])
# ---
# name: test_all_day[empty]
  list([
    tuple(
      'unknown',
      dict({
        'event_type': None,
        'event_types': list([
          'start',
          'end',
        ]),
        'friendly_name': 'Trash day',
        'icon': 'mdi:calendar-text',
      }),
    ),
    tuple(
      '2022-08-08T06:00:00.000+00:00',
      dict({
        'calendar_event': dict({
          'all_day': True,
          'end': '2022-08-09',
          'start': '2022-08-08',
          'summary': 'Trash day calendar event',
        }),
        'event_type': 'start',
        'event_types': list([
          'start',
          'end',
        ]),
        'friendly_name': 'Trash day',
        'icon': 'mdi:calendar-text',
      }),
    ),
    tuple(
      '2022-08-09T06:00:01.000+00:00',
      dict({
        'calendar_event': dict({
          'all_day': True,
          'end': '2022-08-09',
          'start': '2022-08-08',
          'summary': 'Trash day calendar event',
        }),
        'event_type': 'end',
        'event_types': list([
          'start',
          'end',
        ]),
        'friendly_name': 'Trash day',
        'icon': 'mdi:calendar-text',
      }),
    ),
  ])
# ---
# name: test_all_day[lower]
  list([
    tuple(
      'unknown',
      dict({
        'event_type': None,
        'event_types': list([
          'start',
          'end',
        ]),
        'friendly_name': 'Trash day',
        'icon': 'mdi:calendar-text',
      }),
    ),
    tuple(
      '2022-08-08T06:00:00.000+00:00',
      dict({
        'calendar_event': dict({
          'all_day': True,
          'end': '2022-08-09',
          'start': '2022-08-08',
          'summary': 'Trash day calendar event',
        }),
        'event_type': 'start',
        'event_types': list([
          'start',
          'end',
        ]),
        'friendly_name': 'Trash day',
        'icon': 'mdi:calendar-text',
      }),
    ),
    tuple(
      '2022-08-09T06:00:01.000+00:00',
      dict({
        'calendar_event': dict({
          'all_day': True,
          'end': '2022-08-09',
          'start': '2022-08-08',
          'summary': 'Trash day calendar event',
        }),
        'event_type': 'end',
        'event_types': list([
          'start',
          'end',
        ]),
        'friendly_name': 'Trash day',
        'icon': 'mdi:calendar-text',
      }),
    ),
  ])
# ---
# name: test_all_day[options0]
  list([
    tuple(
      'unknown',
      dict({
        'event_type': None,
        'event_types': list([
          'start',
          'end',
        ]),
        'friendly_name': 'Trash day',
        'icon': 'mdi:calendar-text',
      }),
    ),
    tuple(
      '2022-08-08T06:00:00.000+00:00',
      dict({
        'calendar_event': dict({
          'all_day': True,
          'end': '2022-08-09',
          'start': '2022-08-08',
          'summary': 'Trash day calendar event',
        }),
        'event_type': 'start',
        'event_types': list([
          'start',
          'end',
        ]),
        'friendly_name': 'Trash day',
        'icon': 'mdi:calendar-text',
      }),
    ),
    tuple(
      '2022-08-09T06:00:01.000+00:00',
      dict({
        'calendar_event': dict({
          'all_day': True,
          'end': '2022-08-09',
          'start': '2022-08-08',
          'summary': 'Trash day calendar event',
        }),
        'event_type': 'end',
        'event_types': list([
          'start',
          'end',
        ]),
        'friendly_name': 'Trash day',
        'icon': 'mdi:calendar-text',
      }),
    ),
  ])
# ---
# name: test_all_day[options1]
  list([
    tuple(
      'unknown',
      dict({
        'event_type': None,
        'event_types': list([
          'start',
          'end',
        ]),
        'friendly_name': 'Trash day',
        'icon': 'mdi:calendar-text',
      }),
    ),
    tuple(
      '2022-08-08T06:00:00.000+00:00',
      dict({
        'calendar_event': dict({
          'all_day': True,
          'end': '2022-08-09',
          'start': '2022-08-08',
          'summary': 'Trash day calendar event',
        }),
        'event_type': 'start',
        'event_types': list([
          'start',
          'end',
        ]),
        'friendly_name': 'Trash day',
        'icon': 'mdi:calendar-text',
      }),
    ),
    tuple(
      '2022-08-09T06:00:01.000+00:00',
      dict({
        'calendar_event': dict({
          'all_day': True,
          'end': '2022-08-09',
          'start': '2022-08-08',
          'summary': 'Trash day calendar event',
        }),
        'event_type': 'end',
        'event_types': list([
          'start',
          'end',
        ]),
        'friendly_name': 'Trash day',
        'icon': 'mdi:calendar-text',
      }),
    ),
  ])
# ---
# name: test_all_day[options2]
  list([
    tuple(
      'unknown',
      dict({
        'event_type': None,
        'event_types': list([
          'start',
          'end',
        ]),
        'friendly_name': 'Trash day',
        'icon': 'mdi:calendar-text',
      }),
    ),
    tuple(
      '2022-08-08T06:00:00.000+00:00',
      dict({
        'calendar_event': dict({
          'all_day': True,
          'end': '2022-08-09',
          'start': '2022-08-08',
          'summary': 'Trash day calendar event',
        }),
        'event_type': 'start',
        'event_types': list([
          'start',
          'end',
        ]),
        'friendly_name': 'Trash day',
        'icon': 'mdi:calendar-text',
      }),
    ),
    tuple(
      '2022-08-09T06:00:01.000+00:00',
      dict({
        'calendar_event': dict({
          'all_day': True,
          'end': '2022-08-09',
          'start': '2022-08-08',
          'summary': 'Trash day calendar event',
        }),
        'event_type': 'end',
        'event_types': list([
          'start',
          'end',
        ]),
        'friendly_name': 'Trash day',
        'icon': 'mdi:calendar-text',
      }),
    ),
  ])
# ---
# name: test_all_day[upper]
  list([
    tuple(
      'unknown',
      dict({
        'event_type': None,
        'event_types': list([
          'start',
          'end',
        ]),
        'friendly_name': 'Trash day',
        'icon': 'mdi:calendar-text',
      }),
    ),
    tuple(
      '2022-08-08T06:00:00.000+00:00',
      dict({
        'calendar_event': dict({
          'all_day': True,
          'end': '2022-08-09',
          'start': '2022-08-08',
          'summary': 'Trash day calendar event',
        }),
        'event_type': 'start',
        'event_types': list([
          'start',
          'end',
        ]),
        'friendly_name': 'Trash day',
        'icon': 'mdi:calendar-text',
      }),
    ),
    tuple(
      '2022-08-09T06:00:01.000+00:00',
      dict({
        'calendar_event': dict({
          'all_day': True,
          'end': '2022-08-09',
          'start': '2022-08-08',
          'summary': 'Trash day calendar event',
        }),
        'event_type': 'end',
        'event_types': list([
          'start',
          'end',
        ]),
        'friendly_name': 'Trash day',
        'icon': 'mdi:calendar-text',
      }),
    ),
  ])
# ---
# name: test_multiple_all_day
  list([
    tuple(
      '2022-08-08T06:00:00.000+00:00',
      dict({
        'calendar_event': dict({
          'all_day': True,
          'end': '2022-08-09',
          'start': '2022-08-08',
          'summary': 'Trash day calendar event #1',
        }),
        'event_type': 'start',
        'event_types': list([
          'start',
          'end',
        ]),
        'friendly_name': 'Trash day',
        'icon': 'mdi:calendar-text',
      }),
    ),
    tuple(
      '2022-08-09T06:00:01.000+00:00',
      dict({
        'calendar_event': dict({
          'all_day': True,
          'end': '2022-08-09',
          'start': '2022-08-08',
          'summary': 'Trash day calendar event #1',
        }),
        'event_type': 'end',
        'event_types': list([
          'start',
          'end',
        ]),
        'friendly_name': 'Trash day',
        'icon': 'mdi:calendar-text',
      }),
    ),
    tuple(
      '2022-08-09T06:00:01.000+00:00',
      dict({
        'calendar_event': dict({
          'all_day': True,
          'end': '2022-08-09',
          'start': '2022-08-08',
          'summary': 'Trash day calendar event #1',
        }),
        'event_type': 'end',
        'event_types': list([
          'start',
          'end',
        ]),
        'friendly_name': 'Trash day',
        'icon': 'mdi:calendar-text',
      }),
    ),
    tuple(
      '2022-08-15T06:00:00.000+00:00',
      dict({
        'calendar_event': dict({
          'all_day': True,
          'end': '2022-08-16',
          'start': '2022-08-15',
          'summary': 'Trash day calendar event #2',
        }),
        'event_type': 'start',
        'event_types': list([
          'start',
          'end',
        ]),
        'friendly_name': 'Trash day',
        'icon': 'mdi:calendar-text',
      }),
    ),
  ])
# ---
# name: test_overlapping_events
  list([
    tuple(
      'unknown',
      dict({
        'event_type': None,
        'event_types': list([
          'start',
          'end',
        ]),
        'friendly_name': 'Exterior lights',
        'icon': 'mdi:calendar-text',
      }),
    ),
    tuple(
      '2022-08-09T00:30:00.000+00:00',
      dict({
        'calendar_event': dict({
          'all_day': False,
          'end': '2022-08-08T23:00:00-06:00',
          'start': '2022-08-08T18:30:00-06:00',
          'summary': 'Front light',
        }),
        'event_type': 'start',
        'event_types': list([
          'start',
          'end',
        ]),
        'friendly_name': 'Exterior lights',
        'icon': 'mdi:calendar-text',
      }),
    ),
    tuple(
      '2022-08-09T02:30:00.000+00:00',
      dict({
        'calendar_event': dict({
          'all_day': False,
          'end': '2022-08-08T22:00:00-06:00',
          'start': '2022-08-08T20:30:00-06:00',
          'summary': 'Back light',
        }),
        'event_type': 'start',
        'event_types': list([
          'start',
          'end',
        ]),
        'friendly_name': 'Exterior lights',
        'icon': 'mdi:calendar-text',
      }),
    ),
    tuple(
      '2022-08-09T04:00:00.000+00:00',
      dict({
        'calendar_event': dict({
          'all_day': False,
          'end': '2022-08-08T22:00:00-06:00',
          'start': '2022-08-08T20:30:00-06:00',
          'summary': 'Back light',
        }),
        'event_type': 'end',
        'event_types': list([
          'start',
          'end',
        ]),
        'friendly_name': 'Exterior lights',
        'icon': 'mdi:calendar-text',
      }),
    ),
    tuple(
      '2022-08-09T05:00:00.000+00:00',
      dict({
        'calendar_event': dict({
          'all_day': False,
          'end': '2022-08-08T23:00:00-06:00',
          'start': '2022-08-08T18:30:00-06:00',
          'summary': 'Front light',
        }),
        'event_type': 'end',
        'event_types': list([
          'start',
          'end',
        ]),
        'friendly_name': 'Exterior lights',
        'icon': 'mdi:calendar-text',
      }),
    ),
  ])
# ---
